_documentation/Recipes/InlineIPsec.md

```diff
@@ -0,0 +1,63 @@
+   - **Tunnel vs. Transport**: Determines how the original packet is transformed during encryption/decryption as below: 
+![IPsec Packet Transforms](../img/IPsec_Packet_Transforms.svg "IPsec Packet Transforms"){:width="60%"}
+   - **Encrypted Virtual Ports**: Encrypted tunnels between devices over physical ports.  On the wire this wraps additional headers on the outside of the packet in the same way that network virtualization headers such as VXLAN are added/removed.
+   - **Encrypted Physical Ports**: Encrypted virtual ports plugged into compute instances (all traffic going in/out of a given virtual port will be encrypted/decrypted).  On the wire this puts the IPsec headers onto the innermost headers belonging to the end user of the compute instance.
+
+
+The use of cryptography to secure data plane traffic has become common practice. Encrypting and decrypting network traffic using host resources requires that the host must also perform any packet processing of post-decryption receive packets and pre-encryption transmit packets. This can consume significant host compute resources. Offloading packet crypto processing to the data plane removes the burden for cryptographic processing from the host and also allows the infrastructure to be used for pre-encryption and post-decryption packet processing functions.
+
+This recipe looks at inline IPsec processing where the crypto is terminated in the infrastructure, which avoids having specialized drivers or control planes on the compute instance(s). The IPsec control plane (IKE protocol) is offloaded to the infrastructure. The IKE protocol processing runs in software on the infrastructure cores and programs the data plane with offloaded IPsec flows. The data plane programming is separated into two parts, P4Runtime for programming the P4 pipeline and OpenConfig for programming the IPsec crypto configuration. The IPsec crypto configuration includes the IPsec Security Association (SA) table entries (i.e. crypto keys and algorithm selection) and is used for reporting events related to lifetime expiration of SA entries.
+
+# Target Abstraction Interface
+
+Targets supporting this application must export the following tables through the Table-Driven Interface:
+
+1. P4 Tables compiled from the Inline IPsec P4 program.
+* Control blocks for Tunnel and Transport modes.
+* Parser support for IPsec ESP
+* Modify logic to encapsulate on the outer and inner headers of the packet.
+2. IPsec OpenConfig schema based on the RFC
+* Configuration of the type & strength of encryption
+* Authentication mechanisms
+* Classification
+* Access Control Lists
+* Designed to stay aligned with the [OPI Security API specification](https://github.com/opiproject/opi-api/blob/main/security/proto/autogen.md) 
```

Comment:
The OPI Security APIs are meant to sit on top of this. Users would call those APIs to configure IPsec tunnels, the gRPC server would then configure strongSwan via the vici protocol, which would then invoke all of the underlying p4runtime and openconfig things mentioned here. Maybe make this a bit more clear?

Reply:
Added this at the top:  This recipe integrates the [OPI Security API specification](https://github.com/opiproject/opi-api/blob/main/security/proto/autogen.md) to act as a northbound RPC to configure strongSwan remotely.

I didn't mention vici, I wanted to wait until we have this code that I can point to.